File: go-sdk/services/invites_test.go
package services

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strings"
	"testing"

	"github.com/pvium/sdks/go-sdk/config"
	"github.com/pvium/sdks/go-sdk/models"
	"github.com/pvium/sdks/go-sdk/transport"
)

const inviteTestPrivateKey = "0x59c6995e998f97a5a0044976f0d7f3f6f8f53f6a2046baf4f01cb4f1f6bcb58f"

func TestInvitesCreateAndSignOAuthInviteLink(t *testing.T) {
	t.Parallel()

	service := NewInviteService(transport.NewHTTPClient(config.Config{
		BaseURL:     "http://localhost:4005/v1",
		ConsentHost: "http://localhost:3000",
		ClientID:    "app_test",
		APIKey:      "pk_test_dummy",
	}))

	bundle, err := service.CreateBundle(models.OAuthInviteBundleInput{
		Identities: []models.InviteIdentity{{Type: models.InviteIdentityEmail, Value: "Test.User@example.com"}},
		Scopes:     []string{"read:ethereum_wallet", "read:user"},
		Chain:      "ethereum",
	})
	if err != nil {
		t.Fatalf("create bundle: %v", err)
	}
	signed, err := service.SignBundle(bundle, models.OAuthInviteSigner{Chain: "ethereum", PrivateKey: inviteTestPrivateKey})
	if err != nil {
		t.Fatalf("sign bundle: %v", err)
	}

	if signed.ClientID != "app_test" {
		t.Fatalf("client id mismatch: %s", signed.ClientID)
	}
	if len(signed.Invites) != 1 {
		t.Fatalf("expected one invite, got %d", len(signed.Invites))
	}
	if signed.Root.SignatureType != "evm-personal-sign" || signed.Root.SignerAddress == "" {
		t.Fatalf("root signature metadata mismatch: %+v", signed.Root)
	}
	if strings.Join(signed.Scopes, " ") != "read:ethereum_wallet read:user" {
		t.Fatalf("scopes mismatch: %+v", signed.Scopes)
	}

	invite := signed.Invites[0]
	if invite.IdentityType != models.InviteIdentityEmail || invite.IdentityValue != "test.user@example.com" {
		t.Fatalf("identity mismatch: %+v", invite)
	}
	if invite.AppClientID != "app_test" || invite.LeafVersion != "2" || len(invite.SecretHash) != 64 || len(invite.Proof) != 0 {
		t.Fatalf("invite metadata mismatch: %+v", invite)
	}
	inviteURL, err := url.Parse(invite.InviteLink)
	if err != nil {
		t.Fatalf("parse invite link: %v", err)
	}
	query := inviteURL.Query()
	if inviteURL.Scheme+"://"+inviteURL.Host != "http://localhost:3000" || inviteURL.Path != "/oauth2/authorize" {
		t.Fatalf("invite url mismatch: %s", invite.InviteLink)
	}
	if query.Get("client_id") != "app_test" || query.Get("response_type") != "code" {
		t.Fatalf("invite url auth params mismatch: %s", invite.InviteLink)
	}
	if query.Get("scope") != "read:ethereum_wallet read:user" || query.Get("invite_nonce") != invite.InviteNonce || query.Get("invite_secret") != invite.InviteSecret {
		t.Fatalf("invite url invite params mismatch: %s", invite.InviteLink)
	}
	if query.Get("identity_type") != "email" || query.Get("identity_hint") != "test.user@example.com" {
		t.Fatalf("invite url identity params mismatch: %s", invite.InviteLink)
	}
}

func TestInvitesBatchBundleLinksAndCommit(t *testing.T) {
	t.Parallel()

	var gotPath string
	var gotBody map[string]any
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotPath = r.URL.Path
		if err := json.NewDecoder(r.Body).Decode(&gotBody); err != nil {
			t.Fatalf("decode body: %v", err)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"ok":true}`))
	}))
	defer ts.Close()

	service := NewInviteService(transport.NewHTTPClient(config.Config{
		BaseURL:     ts.URL,
		ConsentHost: "http://localhost:3000",
		ClientID:    "app_test",
		APIKey:      "pk_test_dummy",
	}))
	bundle, err := service.CreateBundle(models.OAuthInviteBundleInput{
		Identities: []models.InviteIdentity{{Type: models.InviteIdentityEmail, Value: "Batch.User@example.com"}},
		Scopes:     []string{"read:user", "read:ethereum_wallet"},
		Chain:      "ethereum",
		BatchInvite: &models.OAuthInviteBatchOptions{
			BatchID:     "batch_123",
			StateParams: map[string]any{"source": "sdk-test"},
		},
		StateParams: map[string]any{"returnTo": "/admin/bulk-payments/batch_123"},
	})
	if err != nil {
		t.Fatalf("create bundle: %v", err)
	}
	signed, err := service.SignBundle(bundle, models.OAuthInviteSigner{Chain: "ethereum", PrivateKey: inviteTestPrivateKey})
	if err != nil {
		t.Fatalf("sign bundle: %v", err)
	}
	if signed.BatchID != "batch_123" || signed.BatchInvite.BatchID != "batch_123" {
		t.Fatalf("batch id mismatch: %+v", signed)
	}
	inviteURL, _ := url.Parse(signed.Invites[0].InviteLink)
	if inviteURL.Query().Get("batchId") != "batch_123" {
		t.Fatalf("invite batchId mismatch: %s", signed.Invites[0].InviteLink)
	}
	state, err := url.ParseQuery(inviteURL.Query().Get("state"))
	if err != nil {
		t.Fatalf("parse nested state: %v", err)
	}
	if state.Get("batchId") != "batch_123" || state.Get("source") != "sdk-test" || state.Get("returnTo") != "/admin/bulk-payments/batch_123" {
		t.Fatalf("state mismatch: %s", inviteURL.Query().Get("state"))
	}
	groupURL, _ := url.Parse(signed.GroupInviteLink)
	if groupURL.Query().Get("batchId") != "batch_123" {
		t.Fatalf("group batchId mismatch: %s", signed.GroupInviteLink)
	}

	_, err = service.CommitBundle(context.Background(), signed, nil)
	if err != nil {
		t.Fatalf("commit bundle: %v", err)
	}
	if gotPath != "/batch-payments/batch_123/invites" {
		t.Fatalf("commit path mismatch: %s", gotPath)
	}
	if _, ok := gotBody["root"].(map[string]any); !ok {
		t.Fatalf("commit root missing: %+v", gotBody)
	}
	invites := gotBody["invites"].([]any)
	invite := invites[0].(map[string]any)
	if _, ok := invite["inviteSecret"]; ok {
		t.Fatalf("commit body must omit inviteSecret: %+v", invite)
	}
	if _, ok := invite["inviteLink"]; ok {
		t.Fatalf("commit body must omit inviteLink: %+v", invite)
	}
}

func TestInvitesSupportsSeparateMasterAndRootSigners(t *testing.T) {
	t.Parallel()

	service := NewInviteService(transport.NewHTTPClient(config.Config{
		BaseURL:     "http://localhost:4005/v1",
		ConsentHost: "http://localhost:3000",
		ClientID:    "app_test",
		APIKey:      "pk_test_dummy",
	}))
	bundle, err := service.CreateBundle(models.OAuthInviteBundleInput{
		Identities: []models.InviteIdentity{{Type: models.InviteIdentityEmail, Value: "Split.Signer@example.com"}},
		Scopes:     []string{"read:user", "read:ethereum_wallet"},
		Chain:      "ethereum",
	})
	if err != nil {
		t.Fatalf("create bundle: %v", err)
	}

	calls := []string{}
	signed, err := service.SignBundle(bundle, models.OAuthInviteSigner{
		Chain:         "ethereum",
		SignerAddress: "0x0000000000000000000000000000000000000003",
		SignMessage: func(string) (string, error) {
			t.Fatal("fallback signMessage should not be called")
			return "", nil
		},
		SignMasterSecret: func(message string) (string, error) {
			calls = append(calls, "master:"+message)
			return "0xmaster", nil
		},
		SignInviteRoot: func(message string) (string, error) {
			calls = append(calls, "root:"+message)
			return "0xroot", nil
		},
	})
	if err != nil {
		t.Fatalf("sign bundle: %v", err)
	}
	if len(calls) != 2 || !strings.HasPrefix(calls[0], "master:PVIUM_INVITE_SECRET_V2:") || !strings.HasPrefix(calls[1], "root:PVIUM_INVITE_ROOT_V2") {
		t.Fatalf("unexpected signer calls: %+v", calls)
	}
	if signed.Root.SignerAddress != "0x0000000000000000000000000000000000000003" || signed.Root.Signature != "0xroot" {
		t.Fatalf("root signer mismatch: %+v", signed.Root)
	}
}

func TestCreateBundleRequiresIdentitiesAndBatchIDWhenBatchInviteProvided(t *testing.T) {
	t.Parallel()

	service := NewInviteService(transport.NewHTTPClient(config.Config{
		BaseURL:     "https://api-sandbox.pvium.com/v1",
		ConsentHost: "https://sandbox.pvium.com",
		ClientID:    "client_123",
	}))

	_, err := service.CreateBundle(models.OAuthInviteBundleInput{})
	if err == nil {
		t.Fatal("expected error when identities are empty")
	}
	_, err = service.CreateBundle(models.OAuthInviteBundleInput{
		Identities:  []models.InviteIdentity{{Type: models.InviteIdentityEmail, Value: "payee@example.com"}},
		BatchInvite: &models.OAuthInviteBatchOptions{BatchID: "   "},
	})
	if err == nil {
		t.Fatal("expected error when batchInvite.batchId is empty")
	}
}

File: node-sdk/test/invites.test.js
const test = require("node:test");
const assert = require("node:assert/strict");
const { Wallet, verifyMessage } = require("ethers");

const { PviumSdk } = require("../dist/index.js");

const TEST_PRIVATE_KEY =
  "0x59c6995e998f97a5a0044976f0d7f3f6f8f53f6a2046baf4f01cb4f1f6bcb58f";
const TEST_ADDRESS = new Wallet(TEST_PRIVATE_KEY).address;

test("creates and signs an OAuth invite link with a dummy EVM key", async () => {
  const sdk = PviumSdk.init({
    baseUrl: "http://localhost:4005/v1",
    consentHost: "http://localhost:3000",
    clientId: "app_test",
    apiKey: "pk_test_dummy",
  });

  const bundle = sdk.invites.createBundle({
    identities: [{ type: "email", value: "Test.User@example.com" }],
    scopes: ["read:ethereum_wallet", "read:user"],
    chain: "ethereum",
  });

  const signed = await sdk.invites.signBundle(bundle, {
    chain: "ethereum",
    privateKey: TEST_PRIVATE_KEY,
  });

  assert.equal(signed.clientId, "app_test");
  assert.equal(signed.invites.length, 1);
  assert.equal(signed.root.signatureType, "evm-personal-sign");
  assert.equal(signed.root.signerAddress, TEST_ADDRESS);
  assert.deepEqual(signed.scopes, ["read:ethereum_wallet", "read:user"]);

  const recovered = verifyMessage(
    signed.root.signatureMessage,
    signed.root.signature,
  );
  assert.equal(recovered, TEST_ADDRESS);

  const invite = signed.invites[0];
  assert.equal(invite.identityType, "email");
  assert.equal(invite.identityValue, "test.user@example.com");
  assert.equal(invite.appClientId, "app_test");
  assert.equal(invite.leafVersion, "2");
  assert.equal(typeof invite.secretHash, "string");
  assert.equal(invite.secretHash.length, 64);
  assert.equal(invite.proof.length, 0);

  const inviteUrl = new URL(invite.inviteLink);
  assert.equal(inviteUrl.origin, "http://localhost:3000");
  assert.equal(inviteUrl.pathname, "/oauth2/authorize");
  assert.equal(inviteUrl.searchParams.get("client_id"), "app_test");
  assert.equal(inviteUrl.searchParams.get("response_type"), "code");
  assert.equal(
    inviteUrl.searchParams.get("scope"),
    "read:ethereum_wallet read:user",
  );
  assert.equal(inviteUrl.searchParams.get("invite_nonce"), invite.inviteNonce);
  assert.equal(
    inviteUrl.searchParams.get("invite_secret"),
    invite.inviteSecret,
  );
  assert.equal(inviteUrl.searchParams.get("identity_type"), "email");
  assert.equal(
    inviteUrl.searchParams.get("identity_hint"),
    "test.user@example.com",
  );
});

test("creates batch invite bundle links with explicit batchId and custom state", async () => {
  const requests = [];
  const sdk = PviumSdk.init({
    baseUrl: "http://localhost:4005/v1",
    consentHost: "http://localhost:3000",
    clientId: "app_test",
    apiKey: "pk_test_dummy",
    fetchFn: async (url, init) => {
      requests.push({ url, init });
      const body = JSON.parse(init.body);
      return new Response(JSON.stringify({ data: body.invites }), {
        status: 201,
        headers: { "content-type": "application/json" },
      });
    },
  });

  const bundle = sdk.invites.createBundle({
    identities: [{ type: "email", value: "Batch.User@example.com" }],
    scopes: ["read:user", "read:ethereum_wallet"],
    chain: "ethereum",
    batchInvite: {
      batchId: "batch_123",
      stateParams: { source: "sdk-test" },
    },
    stateParams: { returnTo: "/admin/bulk-payments/batch_123" },
  });

  const signed = await sdk.invites.signBundle(bundle, {
    chain: "ethereum",
    privateKey: TEST_PRIVATE_KEY,
  });

  assert.equal(signed.batchId, "batch_123");
  assert.equal(signed.batchInvite.batchId, "batch_123");

  const inviteUrl = new URL(signed.invites[0].inviteLink);
  assert.equal(inviteUrl.searchParams.get("batchId"), "batch_123");

  const state = new URLSearchParams(inviteUrl.searchParams.get("state"));
  assert.equal(state.get("batchId"), "batch_123");
  assert.equal(state.get("source"), "sdk-test");
  assert.equal(state.get("returnTo"), "/admin/bulk-payments/batch_123");

  const groupUrl = new URL(signed.groupInviteLink);
  assert.equal(groupUrl.searchParams.get("batchId"), "batch_123");

  const committed = await sdk.invites.commitBundle(signed);

  assert.equal(requests.length, 1);
  assert.equal(
    requests[0].url,
    "http://localhost:4005/v1/batch-payments/batch_123/invites",
  );
  assert.equal(committed.inviteCommitted, true);
  assert.equal(committed.alreadyAccepted, false);
  assert.equal(
    committed.committedInvites[0].inviteNonce,
    signed.invites[0].inviteNonce,
  );
});

test("commitBundle detects returned accepted invites with different nonces", async () => {
  const sdk = PviumSdk.init({
    baseUrl: "http://localhost:4005/v1",
    consentHost: "http://localhost:3000",
    clientId: "app_test",
    apiKey: "pk_test_dummy",
    fetchFn: async () =>
      new Response(
        JSON.stringify({
          data: [
            {
              id: "invite_existing",
              identityType: "github",
              identityValue: "feminefa",
              inviteNonce: "existing_nonce",
              status: "accepted",
            },
          ],
        }),
        {
          status: 201,
          headers: { "content-type": "application/json" },
        },
      ),
  });

  const signed = await sdk.invites.createSignedBundle(
    {
      identities: [{ type: "github", value: "feminefa" }],
      scopes: ["read:user"],
      chain: "ethereum",
    },
    {
      chain: "ethereum",
      privateKey: TEST_PRIVATE_KEY,
    },
  );
  const committed = await sdk.invites.commitBundle(signed);

  assert.equal(committed.inviteCommitted, false);
  assert.equal(committed.alreadyAccepted, true);
  assert.equal(committed.existingInvites[0].inviteNonce, "existing_nonce");
  assert.notEqual(
    committed.existingInvites[0].inviteNonce,
    signed.invites[0].inviteNonce,
  );
});

test("supports separate master-secret and invite-root signers", async () => {
  const sdk = PviumSdk.init({
    baseUrl: "http://localhost:4005/v1",
    consentHost: "http://localhost:3000",
    clientId: "app_test",
    apiKey: "pk_test_dummy",
  });
  const wallet = new Wallet(TEST_PRIVATE_KEY);
  const calls = [];

  const bundle = sdk.invites.createBundle({
    identities: [{ type: "email", value: "Split.Signer@example.com" }],
    scopes: ["read:user", "read:ethereum_wallet"],
    chain: "ethereum",
  });

  const signed = await sdk.invites.signBundle(bundle, {
    chain: "ethereum",
    signerAddress: wallet.address,
    signMessage: async () => {
      throw new Error("fallback signMessage should not be called");
    },
    signMasterSecret: async (message) => {
      calls.push(`master:${message}`);
      return wallet.signMessage(message);
    },
    signInviteRoot: async (message) => {
      calls.push(`root:${message}`);
      return wallet.signMessage(message);
    },
  });

  assert.equal(calls.length, 2);
  assert.ok(calls[0].startsWith("master:PVIUM_INVITE_SECRET_V2:"));
  assert.ok(calls[1].startsWith('root:PVIUM_INVITE_ROOT_V2'));
  assert.equal(signed.root.signerAddress, wallet.address);
  assert.equal(
    verifyMessage(signed.root.signatureMessage, signed.root.signature),
    wallet.address,
  );
});

test("finds app invites by identity through the app-invites endpoint", async () => {
  const requests = [];
  const sdk = PviumSdk.init({
    baseUrl: "http://localhost:4005/v1",
    consentHost: "http://localhost:3000",
    clientId: "app_test",
    apiKey: "pk_test_dummy",
    fetchFn: async (url, init) => {
      requests.push({ url, init });
      return new Response(
        JSON.stringify({
          data: [
            {
              id: "invite_1",
              identityType: "github",
              identityValue: "feminefa",
              status: "accepted",
            },
          ],
        }),
        {
          status: 200,
          headers: { "content-type": "application/json" },
        },
      );
    },
  });

  const result = await sdk.invites.findAppInviteByIdentity({
    identityType: "github",
    identityValue: "@FemInefa",
    status: "accepted",
  });

  assert.equal(requests.length, 1);
  const url = new URL(requests[0].url);
  assert.equal(url.pathname, "/v1/batch-payments/app-invites");
  assert.equal(url.searchParams.get("identityType"), "github");
  assert.equal(url.searchParams.get("identityValue"), "@FemInefa");
  assert.equal(url.searchParams.get("status"), "accepted");
  assert.equal(requests[0].init.headers["x-api-key"], "pk_test_dummy");
  assert.equal(result.accepted, true);
  assert.equal(result.invite.id, "invite_1");
});
